File: TEST data for large repo/repo-Copy5/Blueprism/BluePrism.Automate/AutomateAppCore/EnvironmentFunctions/GetStartTimeFunction.vb
Imports BluePrism.AutomateProcessCore

Namespace EnvironmentFunctions

    ''' <summary>
    ''' Function to get the start time of a process instance
    ''' </summary>
    Public Class GetStartTimeFunction : Inherits GetStartTimeFunctionBase
        
        Private ReadOnly mServerFactory As Func(Of IServer)

        Public Sub New(serverFactory As Func(Of IServer))
            mServerFactory = serverFactory
        End Sub

        Protected Overrides Function InnerEvaluate(parameters As IList(Of clsProcessValue), process As clsProcess) As clsProcessValue
            If parameters.Count <> 0 Then
                Throw GetStartTimeFunctionShouldHaveNoParametersException()
            End If

            If process.Session Is Nothing Then
                Throw GetStartTimeFunctionSessionNotRunningException()
            End If

            Dim sessions = mServerFactory().GetActualSessions(process.Session.ID)
            If sessions.Count <> 1 Then
                Throw MissingSessionException()
            End If

            Dim sessionStart = sessions.First().SessionStart.ToUniversalTime().DateTime

            Return New clsProcessValue(DataType.datetime, sessionStart, False)
        End Function
    End Class
End NameSpace

File: TEST data for large repo/repo_Copy-2/Blueprism/BluePrism.DigitalWorker/EnvironmentFunctions/GetStartTimeFunction.cs
using System;
using System.Collections.Generic;
using System.Linq;
using BluePrism.AutomateAppCore;
using BluePrism.AutomateProcessCore;
using BluePrism.DigitalWorker.Sessions;

namespace BluePrism.DigitalWorker.EnvironmentFunctions
{
    public class GetStartTimeFunction : GetStartTimeFunctionBase
    {
        private readonly IRunningSessionRegistry _runningSessionRegistry;

        public GetStartTimeFunction(IRunningSessionRegistry runningSessionRegistry)
        {
            _runningSessionRegistry = runningSessionRegistry ?? throw new ArgumentNullException(nameof(runningSessionRegistry));
        }

        protected override clsProcessValue InnerEvaluate(IList<clsProcessValue> parameters, clsProcess process)
        {
            if (parameters.Any())
                throw GetStartTimeFunctionShouldHaveNoParametersException;

            if (process.Session == null)
                throw MissingSessionException;

            var digitalWorkerRunnerRecord = _runningSessionRegistry.Get(process.Session.ID);

            if (digitalWorkerRunnerRecord == null)
                throw GetStartTimeFunctionSessionNotRunningException;
            
            return new clsProcessValue(DataType.datetime, digitalWorkerRunnerRecord.SessionStarted.ToUniversalTime().DateTime, false);
        }
    }
}
